src/openlcb/TractionThrottle.hxx

```diff
@@ -653,16 +710,109 @@ private:
         }
     }
 
-    /** Allocates (synchronously) an outgoing openlcb buffer with traction
-     * request MTI and the given payload and sends off the message to the bus
-     * for dst_. */
-    void send_traction_message(const Payload &payload)
+    /// Allocates (synchronously) an outgoing openlcb buffer with traction
+    /// request MTI and the given payload and sends off the message to the bus
+    /// for dst_.
+    ///
+    /// Performs loopback to other traction throttles that might be assigned to
+    /// the same train.
+    ///
+    /// @param payload is the data contents of the message
+    /// (e.g. TractionDefs::speed_set_payload(...).
+    void send_traction_message_with_loopback(Payload payload)
+    {
+        auto b = send_traction_message_helper(std::move(payload));
+        std::function<void()> f = std::bind(
+            &TractionThrottle::loopback_traction_message, this, b.release());
+        iface()->executor()->add(new CallbackExecutable(std::move(f)));
+    }
+
+    /// Performs loopback processing of an outgoing traction message. Run on
+    /// the iface()'s executor.
+    ///
+    /// @param b the message that was sent out. Will be unreffed.
+    void loopback_traction_message(Buffer<GenMessage>* b)
+    {
+        auto rb = get_buffer_deleter(b);
+        // Walks all TractionThrottle objects.
+        TractionThrottle *p = nullptr;
+        do
+        {
+            {
+                // finds next instance that's interesting
+                AtomicHolder h(LinkedObject<TractionThrottle>::head_mu());
```

Comment:
I'm not crazy about holding an atomic section for the loop through of an unbounded list. Is there an existing mutex we could use instead? What about a task scheduler lock instead?

Reply:
Under linux an Atomic is actually a mutex.
Under freertos it is a critical section, which means that (on an M3/M4) only priority zero interrupts can go through.
So the important question is how many iterations are there here and how long do they take.
- For almost all devices there are at most two iterations. Throttles typically have one or two such objects.
- For command stations there can be up to 64 such objects. (63 for NCE, 31 for XNet). Definitely not unbounded, because each tractionthrottle is about 100 bytes or more.

One iteration is just a pointer chasing and comparing some memory bytes. I expect it to be say 0.5 to 1 usec (few dozen instructions).
As a result, I expect a non-p0 interrupt jitter of about 0.05 to 0.1 millisecond to be caused by this.

Non-P0 interrutps are things like CAN-bus (1 msec per interrupt) or UART (where we have a FIFO , unless in NCE, where the interrupt is one per msec as well).

It's not super practical to replace this with a mutex or a scheduler lock:
- LinkedObject<> is used elsewhere, if we want to replace the locking, we need to copy the code or impact every other use as well (examples are Executor<> and StateFlowWithQueue<>)
- scheduler lock doesn't exist in linux. We don't have an abstraction for it. It's also unclear what it would do in a multi-core ESP32.
- Another alternative might be to jump to the main executor for sending outgoing packets, but this is also not exciting, because it would probably break all unit tests. It also doesn't guarantee that the links don't change if an object gets constructed or deleted on another thread.